src/main/java/com/google/sps/data/BooksManager.java

```diff
@@ -30,18 +30,71 @@
  * Class used for managing the stock of books. 
  */
 public class BooksManager {
+
+  /**
+   * Returns the Key of book bookId from datastore or null if there is no book with bookId.
+   */
+  private static Key getBookKeyFromDatastore(String bookId) {
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Query bookQuery = new Query("Book")
+        .setFilter(new Query.FilterPredicate("bookId", Query.FilterOperator.EQUAL, bookId))
+        .setKeysOnly();
+    PreparedQuery result = datastore.prepare(bookQuery);
+    if (result.countEntities(FetchOptions.Builder.withDefaults()) == 0) {
+      return null;
+    }
+    return result.asSingleEntity().getKey();
+  }
+
   /** 
    * Returns a List of "Library" Entity objects that have bookId in stock.
    */
-  public List<Entity> getLibrariesForBook(String bookId, boolean setLimit) {
-    throw new UnsupportedOperationException("Method will return a List of LibraryEntities that contain bookId.");
+
+  public List<Entity> getLibrariesForBook(String bookId) {
+    Key bookKey = getBookKeyFromDatastore(bookId);
+    // User can place orders only after the books are checked to be in stock,
+    // so we can consider we only work with books that are in stock somewhere.
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Query libraryQuery = new Query("LibraryStock").setAncestor(bookKey)
+        .setFilter(new Query.FilterPredicate("stock", Query.FilterOperator.NOT_EQUAL, 0));
+    return datastore.prepare(libraryQuery).asList(FetchOptions.Builder.withDefaults());  
   }
 
   /** 
-   * Removes one occurence for each book in bookIds from library. This method is called after an
-   * order for renting books in bookIds from library is added to datastore.  
+   * Removes one occurence for each book in bookIds from library. Also, will remove once
+   * occurence from thte total stock for each book. 
+   * This method is called after an order for renting books in bookIds
+   * from library is added to datastore.  
    */
+
   public void removeBooksFromLibrary(LibraryPoint library, List<String> bookIds) {
-    throw new UnsupportedOperationException("Method will remove bookIds from library in datastore.");
+    // TODO transaction
+    for (String bookId: bookIds) {
+      Key bookKey = getBookKeyFromDatastore(bookId);
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      
+      Query libraryQuery = new Query("LibraryStock").setAncestor(bookKey)
+          .setFilter(new Query.FilterPredicate("libraryId", Query.FilterOperator.EQUAL, (int)library.getLibraryId()));     
+      Entity result = datastore.prepare(libraryQuery).asSingleEntity();
+      // Remove one occurence from library.
+      if (((Number) result.getProperty("stock")).intValue() <= 0) {
+        return;
```

Comment:
How will you know if this succeeded or not?

You need to either return an indicator that this failed (e.g. a boolean) or raise an exception here. Otherwise the caller will expect this to have succeeded, when it didn't.

Please also add tests for what happens when you try to remove books from a library with no stocks. Unit tests are expected to provide coverage for these corner cases.

Reply:
I made this method return a boolean, and in the unsuccessful cases it will return false. Then in placeOrder() method from OrderHandler will check if the method was successful or not, if it returns false, it throws a BadRequestException, since you shouldn't actually ever come to the point where you try to remove from a library with no stocks.
Also addded tests for when the stock is 0.